src/Database/LSMTree/Internal/WriteBufferReader.hs

```diff
@@ -0,0 +1,155 @@
+import           Control.RefCount (Ref, releaseRef)
+import           Control.TempRegistry
+import           Data.Primitive.MutVar (MutVar, newMutVar, readMutVar,
+                     writeMutVar)
+import           Data.Primitive.PrimVar
+import           Data.Word (Word16)
+import           Database.LSMTree.Internal.BlobRef (RawBlobRef (..))
+import qualified Database.LSMTree.Internal.Entry as E
+import           Database.LSMTree.Internal.Lookup (ResolveSerialisedValue)
+import           Database.LSMTree.Internal.Paths
+import qualified Database.LSMTree.Internal.Paths as Paths
+import           Database.LSMTree.Internal.RawPage
+import           Database.LSMTree.Internal.RunReader (Entry (..), Result (..),
+                     mkEntryOverflow, readDiskPage, readOverflowPages,
+                     toFullEntry)
+import           Database.LSMTree.Internal.Serialise (SerialisedValue)
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import           Database.LSMTree.Internal.WriteBufferBlobs (WriteBufferBlobs)
+import qualified Database.LSMTree.Internal.WriteBufferBlobs as WBB
+import qualified System.FS.API as FS
+import           System.FS.API (HasFS)
+import qualified System.FS.BlockIO.API as FS
+import           System.FS.BlockIO.API (HasBlockIO)
+
+{-# SPECIALISE
+    readWriteBuffer ::
+         TempRegistry IO
+      -> ResolveSerialisedValue
+      -> HasFS IO h
+      -> HasBlockIO IO h
+      -> WriteBufferFsPaths
+      -> IO (WriteBuffer, Ref (WriteBufferBlobs IO h))
+  #-}
+readWriteBuffer ::
+     (MonadMVar m, MonadMask m, MonadSTM m, MonadST m)
+  => TempRegistry m
+  -> ResolveSerialisedValue
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> m (WriteBuffer, Ref (WriteBufferBlobs m h))
+readWriteBuffer reg resolve hfs hbio wbPaths =
+  bracket (new reg hfs hbio wbPaths) close $ \reader@WriteBufferReader{..} ->
+    (,) <$> readEntries reader <*> pure readerWriteBufferBlobs
+  where
+    readEntries reader = readEntriesAcc WB.empty
+      where
+        readEntriesAcc acc = next reader >>= \case
+          Empty -> pure acc
+          ReadEntry key entry -> readEntriesAcc $ WB.addEntry resolve key (rawBlobRefSpan <$> toFullEntry entry) acc
+
+-- | Allows reading the k\/ops of a run incrementally, using its own read-only
+-- file handle and in-memory cache of the current disk page.
+--
+-- New pages are loaded when trying to read their first entry.
+data WriteBufferReader m h = WriteBufferReader {
+      -- | The disk page currently being read. If it is 'Nothing', the reader
+      -- is considered closed.
+      readerCurrentPage      :: !(MutVar (PrimState m) (Maybe RawPage))
+      -- | The index of the entry to be returned by the next call to 'next'.
+    , readerCurrentEntryNo   :: !(PrimVar (PrimState m) Word16)
+    , readerKOpsHandle       :: !(FS.Handle h)
+    , readerWriteBufferBlobs :: !(Ref (WriteBufferBlobs m h))
```

Comment:
Could this be a `BlobFile` instead? That would keep this code closer to the original `RunReader`